rtos/Kernel.cpp

```diff
@@ -0,0 +1,52 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2017 ARM Limited
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+#include "rtos/Kernel.h"
+
+#include "cmsis_os2.h"
+
+namespace rtos {
+
+uint64_t Kernel::get_ms_count() {
+    return osKernelGetTickCount();
```

Comment:
I think this was changed back to uint32_t in more recent versions of RTX:
https://github.com/ARM-software/CMSIS_5/commit/9470b0bfa35625542f54afb99ec8ddf70ea8f756#diff-818bf32b54ffc78cfbdfe02740cf27f2R343

Reply:
Gah. Okay, that's going to cause grief. Are we moving to 2.1.1 in the immediate future?

I want this API to be stable - I guess we'd make it 32-bit now, and make sure it's written to cope with both 32-bit and 64-bit underlying?

Also, only 2.1.1 says it's okay to call from ISR, 2.1.0 says it isn't.

Reply:
Brother. Looking deeper, RTX 5.1.1 does actually permit call from interrupt, so could let that pass.

But looking at RTX 5.2.0, I think the implementation (and possibly the CMSIS 2.1.1 spec) is flakey. Wrapping isn't handled sensibly. If the current time is 0x8000 and you say `delay_until(0x7ffe)`, it assumes you want to wait until 2^32-2 ticks in the future, not that you've overrun and are 2 ticks late. That's not a sensible assumption - it would be more conventional to assume half the number space is past and half is future.

2.1.0 is solid, because it returns immediately in that case - it does 64-bit maths, computes "target-now" as 2^64-2, and decides that's not a sensible wait because it is >2^31, and returns osErrorParameter immediately. (Arguably it maybe could have said OK, but it does return immediately, which is the main thing).

I would be very wary about actually using this API with 2.1.1 behaviour. Get CPU starved a bit too long and you're sleeping for 49 days.

Reply:
> Gah. Okay, that's going to cause grief. Are we moving to 2.1.1 in the immediate future?

I would assume we would. There are couple of changes incoming (cortex-a, cortex m33) that might require some updates (some bugs were found and are still being in discussion).

@JonatanAntoni are our findings correct that are described above?

Reply:
Yes, we decided to go back to uint32_t as tick type with CMSIS 5.1 release. osDelayUntil will wait exactly until the given tick time is reached (again). So if you call osDelayUntil with a value slightly lower than osKernelGetTickCount it'll wait 49 days, sure. I think we don't have a chance to predict if you really want to wait 49 days or not. Please let us know if you have reasonable concerns about the current implementation.

Reply:
Just raised an issue over here, going in to detail on my implementation concern: https://github.com/ARM-software/CMSIS_5/issues/277

If I fix this API at 32-bit, and we need to support 2.1.0, I'm going to have to muck around anyway with 32-bit/64-bit adaptation, so it's not going to be straight mapped to osDelayUntil, so maybe I could also muck around in 2.1.1 to adjust the past handling to "31-bit" to suit my taste...

As it happens, I'm not intending to use osDelayUntil at this point anyway, just have an analogue to it in ConditionVariable::wait_until() - its timeout will be based on Semaphore::Wait, so it is free to have its own absolute timestamp interpretation, but behaviour really should be aligned.

Reply:
On the 32-bit/64-bit difference - how about making the timestamp be a typedef, so it matches and adapts to the underlying system? Would be less problematic than trying to handle 32<->64 bit conversions.

Normal users shouldn't care whether it's just 32 or 64 bit, just that the "get ticks" and "until" APIs use the same numbers.

Although I'm not sure how I'd detect it, without grobbling into RTX headers - it's a compile-time difference and there doesn't seem to be a compile-time version macro in "cmsis_os2.h". And we're not using C++11, so no `decltype(osKernelGetTickCount())`.